### PR TITLE
diff(poly) and poly.diff should give same result

### LINKED ISSUE #9585: inconsistent diff on polynomials
From this stackoverflow question: http://stackoverflow.com/questions/31120436/why-does-sympy-diff-not-differentiate-sympy-polynomials-as-expected

``` python
In [1]: from sympy import *

In [2]: from sympy.abc import x

In [3]: p = Poly(x+1, x, domain='QQ')

In [4]: p.diff(x)
Out[4]: Poly(1, x, domain='QQ')

In [5]: diff(p, x)
Out[5]: Derivative(Poly(x + 1, x, domain='QQ'), x)

In [6]: diff(p, x).doit()
Out[6]: Derivative(Poly(x + 1, x, domain='ZZ'), x)
```

Seems like the toplevel `diff` is following a different code path compared to the `diff` method on Poly.

### PR DESCRIPTION
**Pull request by leosartaj**
fixes #9585

### CODE REVIEW

**Comment by leosartaj** [comment]
Polys interface of diff is different from that of sympy's. Therefore unevaluated Derivative creation can fail. eg.
```
>>> p = Poly(x**2 + x, x)
>>> diff(p, (x, 1))
Poly(3*x**2, x, domain='ZZ') # works
>>> diff(p, (x, 1), evaluate=False)
AttributeError
```

This is beacause ``Derivative`` doesn't recognize this interface.

By removing [L-2204](https://github.com/sympy/sympy/pull/9622/files#diff-c96d00473bda7f403c25a06fc5814515R2204) and [L-2205](https://github.com/sympy/sympy/pull/9622/files#diff-c96d00473bda7f403c25a06fc5814515R2205), ``diff`` will always try to evaluate and never return unevaluated ``Derivative``.

These are my thoughts. Suggestions?

**Comment by leosartaj** [comment]
Ping @jksuom @aktech

File: sympy/polys/tests/test_polytools.py [hunk 1]
```diff
@@ -1423,6 +1423,14 @@ def test_Poly_diff():
     assert Poly(x**2*y**2 + x*y).diff(y, x) == Poly(4*x*y + 1)
 
 
+def test_issue_9585():
+    assert diff(Poly(x**2 + x)) == Poly(2*x + 1)
+    assert diff(Poly(x**2 + x), x) == Poly(2*x + 1)
```
Comment (aktech) [inline sympy/polys/tests/test_polytools.py]: This looks redundant, (not related to the PR), isn't it?

**Comment by aktech** [comment]
This looks fine to me, except for the minor [comment](https://github.com/sympy/sympy/pull/9622#discussion_r33917493).

File: sympy/polys/tests/test_polytools.py (hunk 1 above)
Comment (leosartaj) [inline sympy/polys/tests/test_polytools.py]: I will get rid of it.

**Comment by aktech** [comment]
Will commit in 24hr if there is no objection.
ping @pbrady  @jksuom

**Comment by pbrady** [comment]
I'm not too familiar with the Poly implementation but this looks like a good, straightforward fix to me.  Instead of the weird behavior in #9585, I now get:

```python
In [1]: from sympy import *
 
In [2]: from sympy.abc import *

In [3]: p = Poly(x+1, x, domain='QQ')

In [4]: p.diff(x)
Out[4]: Poly(1, x, domain='QQ')

In [5]: diff(p, x)
Out[5]: Poly(1, x, domain='QQ')
```

+1.  Thanks for doing this @leosartaj.

**Comment by aktech** [comment]
@leosartaj  Thanks for working on this. Merging.

**Comment by leosartaj** [comment]
Thanks!

**Pull request merged**